crates/core/src/metadata_store/providers/objstore/optimistic_store.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use rand::Rng;
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::anyhow;
+use arc_swap::ArcSwap;
+use bytestring::ByteString;
+use tokio::time::Instant;
+
+use crate::metadata_store::providers::objstore::immutable_store::ImmutableStore;
+use crate::metadata_store::providers::objstore::version_repository::{
+    VersionRepository, VersionRepositoryError,
+};
+use crate::metadata_store::{MetadataStore, Precondition, ReadError, VersionedValue, WriteError};
+use restate_types::config::MetadataStoreClient;
+use restate_types::Version;
+
+pub(crate) struct OptimisticLockingMetadataStoreBuilder {
+    pub(crate) version_repository: Arc<dyn VersionRepository>,
+    pub(crate) _configuration: MetadataStoreClient,
+}
+
+impl OptimisticLockingMetadataStoreBuilder {
+    pub(crate) async fn build(&self) -> anyhow::Result<OptimisticLockingMetadataStore> {
+        let refresh_duration = Duration::from_secs(5);
+        Ok(OptimisticLockingMetadataStore::new(
+            refresh_duration,
+            self.version_repository.clone(),
+        ))
+    }
+}
+
+struct CachedStore {
+    store: ImmutableStore,
+    created_at: Instant,
+}
+
+impl CachedStore {
+    fn new(store: ImmutableStore) -> Self {
+        Self {
+            store,
+            created_at: Instant::now(),
+        }
+    }
+
+    fn is_stall(&self, duration: Duration) -> bool {
+        self.created_at.elapsed() > duration
+    }
+}
+
+impl Default for CachedStore {
+    fn default() -> Self {
+        Self {
+            store: Default::default(),
+            created_at: Instant::now(),
+        }
+    }
+}
+
+pub struct OptimisticLockingMetadataStore {
+    version_repository: Arc<dyn VersionRepository>,
+    latest_store_cache: ArcSwap<Option<CachedStore>>,
```

Comment:
Can this also be a `CachedStore` (w/o `ArcSwap<Option<>>`) and change the methods to take a mutable borrow?